### PR TITLE
Add instance editor

### PR DESCRIPTION
**Pull request by vadi2**
#### Brief overview of PR changes/additions
Add an instance editor to the errors page which shows up when an issue is clicked upon.
#### Motivation for adding to Hammer
Much easier and faster validate-fix-validate workflow.
#### Other info (issues closed, discussion etc)
If the lines are wrapped, clicking on a message won't take you to the exactly right line. I haven't yet worked out the algorithm for scrolling to the right location in that scenario just yet...

### CODE REVIEW

**Comment by vadi2** [comment]
I'll add the error markings at a later point - it would require moving to a side-by-side instead of the current top-bottom view and adding in a custom numbers column.

@pieter-edelman-nictiz mind reviewing the PR?

**Comment by pieter-edelman-nictiz** [comment]
It doesn't seem to work yet reliably on one of my test instances. I'll try to figure out what's wrong.

**Comment by vadi2** [comment]
OK

**Comment by pieter-edelman-nictiz** [comment]
Right, line wrapping messed with the scrolling. I added a PR to fix this.

File: InstanceEditor.qml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        for (var currentLine = 0; currentLine < lineNumber - 1; currentLine++) {
+            position += lines[currentLine].length + 1 // include newline char
+        }
+        contentItem.contentY = textArea.positionToRectangle(position).y
+    }
+
+    property int selectStart
+    property int selectEnd
+    property int curPos
+
```
Comment (pieter-edelman-nictiz) [inline InstanceEditor.qml]: These don't have to be 'public' properties, you can just make them local variables in the openContextMenu() function below.

File: InstanceEditor.qml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    clip: true
+    ScrollBar.vertical.policy: ScrollBar.AlwaysOn
+    onImplicitHeightChanged: textArea.update()
+
+    TextArea {
+        id: textArea
+        text: myText
```
Comment (pieter-edelman-nictiz) [inline InstanceEditor.qml]: Maybe rename this to just 'text'?

File: Main.qml [hunk 3]
```diff
@@ -126,7 +126,7 @@ ApplicationWindow {
                 id: textArea
                 placeholderText: qsTr("or load it here")
                 renderType: Text.NativeRendering
-                onTextChanged: { appmodel.resourceText = text }
+                onTextChanged: { appmodel.resourceText = text; }
```
Comment (pieter-edelman-nictiz) [inline Main.qml]: (curly braces aren't needed)

File: Main.qml [hunk 4]
```diff
@@ -262,12 +262,15 @@ ApplicationWindow {
 
         Button {
             id: actionButton
+            // this should be set declaratively
```
Comment (pieter-edelman-nictiz) [inline Main.qml]: It isn't?

File: Main.qml [hunk 5]
```diff
@@ -362,21 +342,66 @@ ApplicationWindow {
...
+                    function peekIssue(lineNumber, linePosition) {
+                        if (lineNumber === 0 && linePosition === 0) { return; }
+                        resultsPageEditor.state = "VISIBLE"
+                        resultsPageEditor.scrollToLine(lineNumber)
+                    }
+
+                    // this should be done better with findBlockByLineNumber from QTextDocument(), requires C#
+                    function lineNumberToY(lineNumber) {
+                        return resultsPageEditor.letterHeight * lineNumber;
+                    }
```
Comment (pieter-edelman-nictiz) [inline Main.qml]: This function isn't needed anymore, forgot to remove it

File: InstanceEditor.qml (hunk 2 above)
Comment (vadi2) [reply InstanceEditor.qml]: It seemed to cause issues when I did that - but I'll try again, or give it a better name. This was a crappy temp one

File: InstanceEditor.qml (hunk 1 above)
Comment (vadi2) [reply InstanceEditor.qml]: Applied feedback.

File: Main.qml (hunk 3 above)
Comment (vadi2) [reply Main.qml]: Applied feedback.

File: Main.qml (hunk 4 above)
Comment (vadi2) [reply Main.qml]: Not quite... a few places set it via `PropertyChanges`

File: InstanceEditor.qml (hunk 2 above)
Comment (vadi2) [reply InstanceEditor.qml]: Yes, `text` doesn't work but I gave it a better name.

File: Main.qml (hunk 5 above)
Comment (vadi2) [reply Main.qml]: Removed.

**Comment by vadi2** [comment]
OK, have a look again.

**Pull request merged**